src/Vault.sol

```diff
@@ -0,0 +1,398 @@
+    }
+
+    function _convertAssetToBase(address asset_, uint256 amount) internal view returns (uint256) {
+        uint256 rate = IRateProvider(_getVaultStorage().rateProvider).getRate(asset_);
+        return amount.mulDiv(rate, 10 ** _getAssetStorage().assets[asset_].decimals, Math.Rounding.Floor);
+    }
+
+    function _convertBaseToAsset(address asset_, uint256 baseAmount) internal view returns (uint256) {
+        uint256 rate = IRateProvider(_getVaultStorage().rateProvider).getRate(asset_);
+        return baseAmount.mulDiv(10 ** _getAssetStorage().assets[asset_].decimals, rate, Math.Rounding.Floor);
+    }
+
+    /// @dev Being Multi asset, we need to add the asset param here to deposit the user's asset accordingly.
+    function _deposit(address asset_, address caller, address receiver, uint256 assets, uint256 shares) internal {
+        _getVaultStorage().totalAssets += assets;
+
+        SafeERC20.safeTransferFrom(IERC20(asset_), caller, address(this), assets);
+
+        _mint(receiver, shares);
+
+        emit Deposit(caller, receiver, assets, shares);
+    }
+
+    // QUESTION: How might we
+    function _withdraw(address caller, address receiver, address owner, uint256 assets_, uint256 shares) internal {
+        // reduce the totalAssets QUESTION: Should this go in the deposit function:
+        _getVaultStorage().totalAssets -= assets_;
+
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        // Check if the contract has enough balance
+        uint256 contractBalance = IERC20(asset()).balanceOf(address(this));
+        if (contractBalance < assets_) {
+            uint256 shortfall = assets_ - contractBalance;
+
+            // pull base asset shortfall from the buffer strategy
+            IStrategy(_getVaultStorage().bufferStrategy).withdraw(shortfall, address(this), address(this));
+        }
+
+        _burn(owner, shares);
+        SafeERC20.safeTransfer(IERC20(asset()), receiver, assets_);
+
+        emit Withdraw(caller, receiver, owner, assets_, shares);
+    }
+
+    function _decimalsOffset() internal pure returns (uint8) {
+        return 0;
+    }
+
+    function _getVaultStorage() internal pure returns (VaultStorage storage $) {
+        assembly {
+            $.slot := 0x22cdba5640455d74cb7564fb236bbbbaf66b93a0cc1bd221f1ee2a6b2d0a2427
+        }
+    }
+
+    function _getAssetStorage() internal pure returns (AssetStorage storage $) {
+        assembly {
+            $.slot := 0x2dd192a2474c87efcf5ffda906a4b4f8a678b0e41f9245666251cfed8041e680
+        }
+    }
+
+    function _getStrategyStorage() internal pure returns (StrategyStorage storage $) {
+        assembly {
+            $.slot := 0x36e313fea70c5f83d23dd12fc41865566e392cbac4c21baf7972d39f7af1774d
+        }
+    }
+
+    //// ADMIN ////
+
+    function setRateProvider(address rateProvider_) external onlyRole(DEFAULT_ADMIN_ROLE) {
```

Comment:
can be done at the end but i'd keep DEFAULT_ADMIN_ROLE only as a role manager,

and have decoupled roles for these set calls, to allow flexibiliy

Reply:
yeah good feedback, I'm going to leave the roles to the end once I see how everything shapes up